### PR TITLE
ui: timeline component added to details pages

### LINKED ISSUE #2775: ui: add timeline component to details pages
In the current UI, we have a few places where we can see the connections between a build - deployment - release, but these are not prominent. Users may want to find which build was used for this deployment, etc, and having these links clearly displayed in the UI will help.

Thus, we would like to add a timeline component to each artifact's details page to show clearly the succession of events.

(see design documents for actual UI)

### PR DESCRIPTION
**Pull request by sabrinako**
## Notes
Currently does not show destroyed deployments in the timeline, @almonk is still working on designs for the best way to show this

## How to test
1. Pull down the branch `git checkout ui/timeline-component`
2. `yarn start`
3. Check the build, deployment, and release details pages for the timelines
_Optional_
4. Spin up a Waypoint server + make some builds, deployments, and releases
5. `ember serve local`
6. Check the build, deployment, and release details pages for the timelines

## Screenshot
Screen Shot 2021-12-14 at 9 28 35 AM

### CODE REVIEW

File: ui/app/components/timeline.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+export default class Timeline extends Component<TimelineArgs> {
+  @service api!: ApiService;
+  @service router!: RouterService;
+
+  get artifacts(): Artifact[] {
+    let timelineObjects = [] as Artifact[];
+    let preload = this.args.currentArtifact.preload;
+    let build = preload?.build ?? null;
+    let deployment = preload?.deployment ?? null;
```
Comment (jgwhite) [inline ui/app/components/timeline.ts:39]: To get around the TypeScript errors here, I’d take a look at [type predicates](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates). Really powerful/useful feature of TypeScript.

**Comment by jgwhite** [comment]
@sabrinako regarding the wonky timestamps, a couple things to keep in mind:

1. Do we show `status.startTime` or `status.completeTime`? Does it depend on the current state of the operation?
2. We do some monkey business to treat a Build and its PushedArtifact as one thing. You’ll see evidence of this in the template for the build page. Do we need to account for that too?

File: ui/app/routes/workspace/projects/project/app/deployment/deployment-seq.ts [hunk 2]
```diff
@@ -32,16 +37,46 @@ export default class DeploymentDetail extends Route {
...
+      timeline.release = {
+        sequence: release.sequence,
+        status: release.status,
+      };
+    }
+
+    return { ...deployment, release, timeline };
+  }
+
+  resetController(_: Controller, isExiting: boolean, transition: Transition): void {
```
Comment (sabrinako) [inline ui/app/routes/workspace/projects/project/app/deployment/deployment-seq.ts:74]: Fixes a bug if you clicked on the 'deployments' tab when already being there, you get redirected to a blank page.

**Comment by jgwhite** [comment]
This feedback might be a little wacky, but I wonder if the timeline should run in the other direction, just so it matches the nav on the left?

As it is, they run in opposite directions:

CleanShot 2021-12-14 at 16 41 10@2x

**Comment by sabrinako** [comment]
> This feedback might be a little wacky, but I wonder if the timeline should run in the other direction, just so it matches the nav on the left?
> 
> As it is, they run in opposite directions:
> 
> CleanShot 2021-12-14 at 16 41 10@2x

I thought about that as well, but I went with how it's structured in the Figma. @almonk any thoughts?

File: ui/app/routes/workspace/projects/project/app/build.ts [hunk 3]
```diff
@@ -34,8 +39,15 @@ export default class BuildDetail extends Route {
...
 
   async model(params: Params): Promise<Model> {
-    let { builds } = this.modelFor('workspace.projects.project.app') as AppRouteModel;
+    let { builds, deployments, releases } = this.modelFor('workspace.projects.project.app') as AppRouteModel;
     let buildFromAppRoute = builds.find((obj) => obj.sequence === Number(params.sequence));
+    let deploymentFromAppRoute = deployments.find((obj) => {
+      if (obj.preload && obj.preload.artifact) {
+        return obj.preload.artifact.id === buildFromAppRoute?.pushedArtifact?.id;
+      }
+      return obj.artifactId === buildFromAppRoute?.pushedArtifact?.id;
```
Comment (jgwhite) [inline ui/app/routes/workspace/projects/project/app/build.ts:48]: I suspect `preload.artifact.id` and `artifactId` will always concur when working with the real API. Maybe something we can fix in Mirage? What do you think? Worth digging into it?

File: ui/app/routes/workspace/projects/project/app/build.ts [hunk 4]
```diff
@@ -51,6 +63,28 @@ export default class BuildDetail extends Route {
...
+
+    if (releaseFromAppRoute) {
+      timeline.release = {
+        sequence: releaseFromAppRoute.sequence,
+        status: releaseFromAppRoute.status,
+      };
+    }
+
+    result.timeline = timeline;
+
```
Comment (jgwhite) [inline ui/app/routes/workspace/projects/project/app/build.ts:87]: Note for the future: we could potentially do this cross-referencing server-side in [the UI_ methods](https://github.com/hashicorp/waypoint/blob/0ee3f72cee308cd9d3a14b5691f324bb69099571/internal/server/proto/server.proto#L369-L376).

File: ui/app/routes/workspace/projects/project/app/deployment/deployment-seq.ts (hunk 2 above)
Comment (jgwhite) [reply ui/app/routes/workspace/projects/project/app/deployment/deployment-seq.ts:74]: Good catch!

File: ui/app/services/api.ts [hunk 5]
```diff
@@ -46,6 +46,11 @@ const protocolVersions = {
   'client-version': 'ui-0.0.1',
 };
 
+export type BuildExtended = Build.AsObject & {
+  statusReport?: StatusReport.AsObject;
+  pushedArtifact?: PushedArtifact.AsObject;
+};
+
```
Comment (jgwhite) [inline ui/app/services/api.ts:53]: Did we end up using this anywhere?

File: ui/app/styles/_variables.scss [hunk 6]
```diff
@@ -24,6 +24,8 @@ $button-shadow: 0 3px 2px rgba(var(--shadow), 0.2);
   --card: #{dehex(color.$white)};
   --highlighted-list-item: #{dehex(color.$ui-cool-gray-100)};
   --panel: #{dehex(color.$ui-cool-gray-050)};
+  --timeline-dot: #{dehex(color.$ui-gray-400)};
+  --timeline-dash: #{dehex(color.$ui-gray-100)};
```
Comment (jgwhite) [inline ui/app/styles/_variables.scss:28]: Do we need dark mode equivalents of these? (Can wait until the polishing phase, of course 😎)

File: ui/app/styles/components/image-ref.scss [hunk 7]
```diff
@@ -4,6 +4,7 @@
 
   &__tag {
     margin-left: 7px;
+    overflow-wrap: anywhere;
```
Comment (jgwhite) [inline ui/app/styles/components/image-ref.scss:7]: Didn’t realize `anywhere` was a possible value here. TIL!

File: ui/app/styles/components/timeline.scss [hunk 8]
```diff
@@ -0,0 +1,58 @@
+.timeline {
+  li {
+    display: grid;
+    grid-template-columns: scale.$sm-1 auto;
+    grid-template-rows: max-content 20px;
```
Comment (jgwhite) [inline ui/app/styles/components/timeline.scss:5]: Loving this use of grid (although I must admit I don’t understand it yet!)

**Review by jgwhite — APPROVED** [review]
This is looking *wonderful*. I only have superficial comments.

Following this PR, we should discuss as a team how we can make our routes a bit less busy. The `afterModel` hooks are a totally logical place to cross-reference and prepare model data, but they are getting a little large and complex. I’m not sure what the alternative is, but together I’m sure we can dream up some neat ideas.

File: ui/app/routes/workspace/projects/project/app/build.ts (hunk 3 above)
Comment (sabrinako) [reply ui/app/routes/workspace/projects/project/app/build.ts:49]: As discussed on sync, this will be a TODO and handled later since it works as-is

**Pull request merged**